File: core/java/src/test/java/hara/spec/SpecRegistry.java
package hara.spec;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;

/** Resolves the workspace's canonical hara-specs-registry checkout for tests. */
public final class SpecRegistry {
  private static final String PROPERTY = "hara.specs.registry";
  private static final String REGISTRY_ENV = "HARA_SPECS_REGISTRY";
  private static final String WORKSPACE_ENV = "HARA_WORKSPACE_ROOT";

  private SpecRegistry() {}

  /** Returns the validated absolute registry root. */
  public static Path root() {
    String configured = System.getProperty(PROPERTY);
    if (configured != null && !configured.isBlank()) {
      return validate(Path.of(configured), "system property -D" + PROPERTY);
    }

    configured = System.getenv(REGISTRY_ENV);
    if (configured != null && !configured.isBlank()) {
      return validate(Path.of(configured), REGISTRY_ENV);
    }

    String workspace = System.getenv(WORKSPACE_ENV);
    if (workspace != null && !workspace.isBlank()) {
      return validate(Path.of(workspace).resolve("technology/hara-specs-registry"), WORKSPACE_ENV);
    }

    Path start = Path.of(System.getProperty("user.dir", ".")).toAbsolutePath().normalize();
    for (Path cursor = start; cursor != null; cursor = cursor.getParent()) {
      for (Path candidate :
          List.of(cursor.resolve("hara-specs-registry"), cursor.resolve("technology/hara-specs-registry"))) {
        if (isRegistry(candidate)) return candidate.toAbsolutePath().normalize();
      }
    }

    throw new IllegalStateException(
        "Cannot locate hara-specs-registry. Set "
            + REGISTRY_ENV
            + " or "
            + WORKSPACE_ENV
            + ", or pass -D"
            + PROPERTY
            + "=/absolute/path/to/hara-specs-registry; working directory was "
            + start);
  }

  /** Returns whether the canonical registry can be found without throwing. */
  public static boolean available() {
    try {
      root();
      return true;
    } catch (IllegalStateException ex) {
      return false;
    }
  }

  /** Resolves a registry-relative file and fails with a useful diagnostic if it is absent. */
  public static Path require(String relative) {
    Path path = resolve(relative);
    if (!Files.isRegularFile(path)) {
      throw new IllegalStateException("Missing specification file: " + path);
    }
    return path;
  }

  /** Resolves a registry-relative file without requiring it to exist. */
  public static Path resolve(String relative) {
    Path requested = Path.of(relative);
    if (requested.isAbsolute() || relative.startsWith("..")) {
      throw new IllegalArgumentException("Specification path must be registry-relative: " + relative);
    }
    Path root = root();
    Path path = root.resolve(requested).normalize();
    if (!path.startsWith(root)) {
      throw new IllegalArgumentException("Specification path escapes the registry: " + relative);
    }
    return path;
  }

  private static Path validate(Path candidate, String source) {
    Path root = candidate.toAbsolutePath().normalize();
    if (!isRegistry(root)) {
      throw new IllegalStateException(
          "Configured hara-specs-registry from " + source + " is not valid: " + root);
    }
    return root;
  }

  private static boolean isRegistry(Path candidate) {
    return Files.isDirectory(candidate)
        && (Files.isRegularFile(candidate.resolve("spec-manifest.json"))
            || Files.isRegularFile(candidate.resolve("registry-index.json")));
  }
}

File: core/rust/src/spec_registry.rs
//! Workspace-independent access to the canonical `hara-specs-registry` checkout.
//!
//! Registry consumers may run from Cargo, Maven, an editor, or a packaged
//! checkout. They must not encode one caller's working-directory layout.

use std::env;
use std::path::{Path, PathBuf};

const REGISTRY_ENV: &str = "HARA_SPECS_REGISTRY";
const WORKSPACE_ENV: &str = "HARA_WORKSPACE_ROOT";

/// Finds the canonical registry using explicit configuration before workspace
/// and ancestor discovery.
pub fn root() -> Option<PathBuf> {
    if let Some(configured) = nonempty_var(REGISTRY_ENV) {
        return valid_root(PathBuf::from(configured));
    }

    if let Some(workspace) = nonempty_var(WORKSPACE_ENV) {
        return valid_root(PathBuf::from(workspace).join("technology/hara-specs-registry"));
    }

    let mut starts = Vec::new();
    if let Ok(current) = env::current_dir() {
        starts.push(current);
    }
    for start in starts {
        for cursor in start.ancestors() {
            for candidate in [
                cursor.join("hara-specs-registry"),
                cursor.join("technology/hara-specs-registry"),
            ] {
                if let Some(root) = valid_root(candidate) {
                    return Some(root);
                }
            }
        }
    }
    None
}

/// Resolves a registry-relative path when the registry is available.
pub fn resolve(relative: &str) -> Option<PathBuf> {
    let requested = Path::new(relative);
    if requested.is_absolute()
        || requested
            .components()
            .any(|component| matches!(component, std::path::Component::ParentDir))
    {
        return None;
    }
    root().map(|registry| registry.join(requested))
}

/// Resolves a required registry path and reports the configuration that was
/// missing instead of exposing a caller-specific relative path failure.
pub fn require(relative: &str) -> PathBuf {
    let path = resolve(relative).unwrap_or_else(|| {
        panic!(
            "cannot locate hara-specs-registry for `{relative}`; set {REGISTRY_ENV} or {WORKSPACE_ENV}"
        )
    });
    assert!(
        path.is_file(),
        "missing hara-specs-registry file: {}",
        path.display()
    );
    path
}

fn nonempty_var(name: &str) -> Option<String> {
    env::var(name).ok().filter(|value| !value.trim().is_empty())
}

fn valid_root(candidate: PathBuf) -> Option<PathBuf> {
    let root = candidate.canonicalize().ok()?;
    let has_manifest = root.join("spec-manifest.json").is_file();
    let has_index = root.join("registry-index.json").is_file();
    (root.is_dir() && (has_manifest || has_index)).then_some(root)
}
